sql/src/main/java/io/crate/executor/transport/ChainableActions.java

```diff
@@ -72,23 +75,59 @@
             CompletableFuture<R> previousActionUndo = previousActions.get(previousActionsSize - 1).undo();
             for (int i = previousActionsSize - 2; i >= 0; i--) {
                 ChainableAction<R> previousAction = previousActions.get(i);
-                previousActionUndo = previousActionUndo.thenCompose(r -> previousAction.undo());
+                previousActionUndo = previousActionUndo
+                    .handle((r, f) -> Result.mergeWith(result, r, f))
+                    .thenCompose(r -> {
+                        if (r.errorCount > 1) {
+                            // last undo also throws an exception. throw it, will stop undo
+                            Exceptions.rethrowUnchecked(r.error);
+                        }
+                        return previousAction.undo();
+                    });
             }
-            return previousActionUndo.whenComplete((r, f) -> Exceptions.rethrowUnchecked(result.error));
+            return previousActionUndo
+                .handle((r, f) -> Result.mergeWith(result, r, f))
+                .thenCompose(r -> {
+                    Exceptions.rethrowUnchecked(result.error);
+                    return CompletableFutures.failedFuture(result.error);
+                });
         }
         return CompletableFuture.completedFuture(result.result);
     }
 
     private static class Result<R> {
 
+        static <R> Result<R> mergeWith(Result<R> r, @Nullable R result, @Nullable Throwable error) {
+            if (result != null) {
+                r.result = result;
+            }
+            if (error != null) {
+                r.addError(error);
+            }
+            return r;
+        }
+
         @Nullable
-        private final R result;
+        private R result;
         @Nullable
-        private final Throwable error;
+        private Throwable error;
+        private int errorCount = 0;
 
         public Result(@Nullable R result, @Nullable Throwable error) {
             this.result = result;
             this.error = error;
+            if (error != null) {
+                errorCount++;
+            }
+        }
+
+        void addError(Throwable t) {
+            if (error != null) {
+                error = new MultiException(ImmutableList.of(error, t));
```

Comment:
If a second error happens, it must happened while undo. we will stop here and throw the error , no further undo actions are executed. So either the error thrown is a single exception or a MultiException containing 2 single exceptions. refactored the code a bit to make this more clear.